### PR TITLE
Events and logging

### LINKED ISSUE #25: Solidity feature: Logging and Events

### PR DESCRIPTION
**Pull request by dfilaretti**
Fixes #25.
 
This introduces the basic machinery for events & logging. 
Notice that not all cases are supported, since for full support we also need #80.

Do not review yet.

### CODE REVIEW

**Comment by dfilaretti** [comment]
Jenkins test this please

**Comment by dfilaretti** [comment]
test this please

**Comment by dfilaretti** [comment]
Jenkins: test this please

**Comment by dfilaretti** [comment]
The following tests were previously unable to compile, while now they compile but fail for other reasons:

## Exception `std::exception: not implemented yet`

* `event_anonymous_with_topics` (string literal)
* `event_really_lots_of_data_from_storage` (array)
* `event_really_really_lots_of_data_from_storage` (array)
* `event_indexed_string` (event)

## Exception `std::exception: not implemented yet: cryptographic functions`

* `event`

## Exception `IeleCompiler: member not supported in IELE`

* `event_really_lots_of_data` (data)

**Comment by dfilaretti** [comment]
The following fails are because they have events with more than one non-indexed argument, which we do not support yet:

* `events_with_same_name`
* `events_with_same_name_inherited`
* `events_with_same_name_inherited_emit`
* `event_lots_of_data`

These should pass once we implement #80

**Comment by dfilaretti** [comment]
Finally, the following represent a genuine issue with the current event/logging machinery, which I am investigating:

* `event_access_through_base_name_emit`
* `event_access_through_base_name`
* `event_no_arguments`
* `event_emit`

**Comment by dfilaretti** [comment]
The 4 failures identifier before are now fixed.

**Comment by dfilaretti** [comment]
Ready for review!

File: libiele/IeleInstruction.cpp [hunk 1]
```diff
@@ -82,6 +110,7 @@ IeleInstruction *IeleInstruction::CreateRet(
                                        ReturnValues.end());
 
   return RetInst;
+
```
Comment (theo25) [inline libiele/IeleInstruction.cpp]: nit: remove this newline

File: libiele/IeleIntConstant.cpp [hunk 2]
```diff
@@ -5,34 +5,42 @@
...
   std::string Indent(indent, ' ');
-  OS << Indent << Val.str();
+  if (PrintAsHex) {
+    OS << boost::str(boost::format("%1$#x") % Val);
+  } else {
+    OS << Indent << Val.str();
+  }
 }
+
+// format(Val) % x1 % x2
```
Comment (theo25) [inline libiele/IeleIntConstant.cpp]: nit: remove comment

File: libsolidity/codegen/IeleCompiler.cpp [hunk 3]
```diff
@@ -1285,6 +1287,98 @@ bool IeleCompiler::visit(const FunctionCall &functionCall) {
...
+    iele::IeleInstruction::CreateBinOp(iele::IeleInstruction::Add,
+                                       NextFree,
+                                       LastUsed, 
+                                       iele::IeleIntConstant::getOne(&Context),
+                                       CompilingBlock);
+
+    // i.e. store %next.free, 0 (needed?)
+    // iele::IeleInstruction::CreateStore(NextFree,
+    //                                    iele::IeleIntConstant::getZero(&Context), 
+    //                                    CompilingBlock);
```
Comment (theo25) [inline libsolidity/codegen/IeleCompiler.cpp]: you shouldn't need this because all memory is automatically initialized to zero

**Pull request merged**